pdfgen.c

```diff
@@ -2292,6 +2304,661 @@ static int pdf_add_barcode_39(struct pdf_doc *pdf, struct pdf_object *page,
+        case '1':
+        case '2':
+            if ((set == 0 && bar) || (set != 0 && !bar)) {
+                width -= 1.0 / 13.0;
+            } else {
+                width += 1.0 / 13.0;
+            }
+            break;
+
+        case '7':
+        case '8':
+            if ((set == 0 && bar) || (set != 0 && !bar)) {
+                width += 1.0 / 13.0;
+            } else {
+                width -= 1.0 / 13.0;
+            }
+            break;
+        }
+
+        width *= x_width;
+        if (bar) {
+            if (pdf_add_filled_rectangle(pdf, page, x, y, width, height, 0,
+                                         colour, PDF_TRANSPARENT) < 0)
+                return pdf->errval;
+        }
+        x += width;
+    }
+    if (new_x)
+        *new_x = x;
+    return 0;
+}
+
+static int pdf_barcode_eanupc_aux(struct pdf_doc *pdf,
+                                  struct pdf_object *page, float x, float y,
+                                  float x_width, float height,
+                                  uint32_t colour, int guard_type,
+                                  float *new_x)
+{
+    int code = code_eanupc_aux_encoding[guard_type];
+
+    for (int i = 5; i >= 0; i--) {
+        int value = code >> i * 2 & 0x3;
+        if (value) {
+            if ((i & 0x1) == 0) {
+                if (pdf_add_filled_rectangle(pdf, page, x, y, x_width * value,
+                                             height, 0, colour,
+                                             PDF_TRANSPARENT) < 0)
+                    return pdf->errval;
+            }
+            x += x_width * value;
+        }
+    }
+    if (new_x)
+        *new_x = x;
+    return 0;
+}
+
+static int pdf_add_barcode_ean13(struct pdf_doc *pdf, struct pdf_object *page,
+                                 float x, float y, float width, float height,
+                                 const char *string, uint32_t colour)
+{
+    if (!string)
+        return 0;
+
+    size_t len = strlen(string);
+    int lead = 0;
+    if (len == 13) {
+        char ch = string[0];
+        if (!isdigit(*string))
+            return pdf_set_err(pdf, -EINVAL,
+                               "Invalid EAN13 character %c 0x%x", ch, ch);
+        lead = ch - '0';
+        ++string;
+    } else if (len != 12)
+        return pdf_set_err(pdf, -EINVAL, "Invalid EAN13 string length %lu",
+                           len);
+
+    /* Scale and calculate dimensions */
+    float x_off, y_off, new_width, new_height, x_width, bar_height, bar_ext,
+        font;
+
+    pdf_barcode_eanupc_calc_dims(PDF_BARCODE_EAN13, width, height, &x_off,
+                                 &y_off, &new_width, &new_height, &x_width,
+                                 &bar_height, &bar_ext, &font);
+
+    x += x_off;
+    y += y_off;
+    float bar_y = y + new_height - bar_height;
+
+    int e;
+    const char *save_font = pdf->current_font->font.name;
+    e = pdf_set_font(pdf, "Courier"); /* Built-in monospace font */
```

Comment:
Should we force the font here, or leave it with what the user has previously set? Overriding it seems a bit odd.

Reply:
The GS1 standard section 5.2.5 on "Human readable interpretation" talks about the text digits representation that goes along with a barcode. It says that the font used should be a clearly legible font, and should fulfill the features defined by ISO 1073-2 and/or the OCR-B font. Since we can't be sure what font the user previously had set it seemed safest to set it to a known in-built monospace font, which will be similar to OCR-B, so I chose Courier.